Wrappers/Python/ccpi/viewer/utils/hdf5_io.py

```diff
@@ -55,64 +58,83 @@ def __init__(self):
+
+    def _update_output_data(self, outInfo):
+        if self._DatasetName is None:
             raise Exception("DataSetName must be set.")
-        with h5py.File(self.__FileName, 'r') as f:
+        with h5py.File(self._FileName, 'r') as f:
             info = outInfo.GetInformationObject(0)
-            shape = np.shape(f[self.__DatasetName])
+            shape = np.shape(f[self._DatasetName])
             # print("keys:", list(f.keys()))
             # print(shape)
             ue = info.Get(vtk.vtkStreamingDemandDrivenPipeline.UPDATE_EXTENT())
             # Note that we flip the update extents because VTK is Fortran order
             # whereas h5py reads in C order. When writing we pretend that the
             # data was C order so we have to flip the extents/dimensions.
             if len(shape) == 3:
-                data = f[self.__DatasetName][ue[4]:ue[5]+1, ue[2]:ue[3]+1, ue[0]:ue[1]+1]
+                data = f[self._DatasetName][ue[4]:ue[5]+1, ue[2]:ue[3]+1, ue[0]:ue[1]+1]
             elif len(shape) == 4:
-                data = f[self.__DatasetName][self.__4DIndex][
+                data = f[self._DatasetName][self._4DIndex][
                     ue[4]:ue[5] + 1, ue[2]:ue[3]+1, ue[0]:ue[1]+1]
             # print("attributes: ", f.attrs.items())
             output = dsa.WrapDataObject(vtk.vtkImageData.GetData(outInfo))
             output.SetExtent(ue)
-            output.PointData.append(data.ravel(), self.__DatasetName)
-            output.PointData.SetActiveScalars(self.__DatasetName)
-            return 1
+            output.PointData.append(data.ravel(), self._DatasetName)
+            output.PointData.SetActiveScalars(self._DatasetName)
+            return output
+        
 
     def SetFileName(self, fname):
-        if fname != self.__FileName:
+        if fname != self._FileName:
             self.Modified()
-            self.__FileName = fname
+            if self._DatasetName is not None:
+                with h5py.File(fname, 'r') as f:
+                    if not (self._DatasetName in f):
+                        raise Exception("No dataset named {} exists in {}.".format(self._DatasetName, fname))
+            self._FileName = fname
 
     def GetFileName(self):
-        return self.__FileName
+        return self._FileName
 
     def SetDatasetName(self, lname):
-        if lname != self.__DatasetName:
+        if lname != self._DatasetName:
             self.Modified()
-            self.__DatasetName = lname
+            if self._FileName is not None:
+                with h5py.File(self._FileName, 'r') as f:
+                    if not (lname in f):
+                        raise Exception("No dataset named {} exists in {}.".format(lname, self._FileName))
+            self._DatasetName = lname
 
     def GetDatasetName(self):
-        return self.__DatasetName
+        return self._DatasetName
 
     def Set4DIndex(self, index):
         '''Sets which index to read, in the case of a 4D dataset'''
```

Comment:
Does this mean that this will only return a 3D object?
I find this index a bit confusing. Is it used to tell which index not to read?

Reply:
yes, only a 3D object is returned, so it's selecting which index in the 4th dimension to read

Reply:
To make it ~more complicated~ clearer, I'd use the complementary information, i.e. request what information we actually need/want.

Reply:
Sorry what do you mean by this?